drivers/power_domain/power_domain_gpio_monitor.c

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright (c) 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT power_domain_gpio_monitor
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(power_domain_gpio_monitor, CONFIG_POWER_DOMAIN_LOG_LEVEL);
+
+struct pd_gpio_monitor_config {
+	struct gpio_dt_spec power_good_gpio;
+};
+
+struct pd_gpio_monitor_data {
+	struct gpio_callback callback;
+	const struct device *dev;
+};
+
+static int pd_gpio_monitor_update_status(struct pd_gpio_monitor_data *data)
+{
+	const struct pd_gpio_monitor_config *config = data->dev->config;
+	int rc;
+
+	rc = gpio_pin_get_dt(&config->power_good_gpio);
+	if (rc < 0) {
+		LOG_WRN("Failed to read gpio logic level");
+		return rc;
+	}
+
+	if (rc == 0) {
+		pm_device_children_action_run(data->dev, PM_DEVICE_ACTION_TURN_OFF, NULL);
+		return -EAGAIN;
+	}
+
+	pm_device_children_action_run(data->dev, PM_DEVICE_ACTION_TURN_ON, NULL);
+	return 0;
+}
+
+static void pd_gpio_monitor_callback(const struct device *port,
+				    struct gpio_callback *cb, gpio_port_pins_t pins)
+{
+	struct pd_gpio_monitor_data *data = CONTAINER_OF(cb, struct pd_gpio_monitor_data, callback);
+
+	(void)pd_gpio_monitor_update_status(data);
+
+}
+
+static int pd_gpio_monitor_pm_action(const struct device *dev, enum pm_device_action action)
```

Comment:
I want to support runtime PM. Every time `pm_device_runtime_get` is called, first power domain gets PM_DEVICE_ACTION_RESUME and then device can be resumed - only if ACTION_RESUME was successful on power domain.

Reply:
The goal of a power domain is for dependent drivers to be configured appropriately for the power state they are physically in. If we consider a device on this power domain, where application code calls:
```
pm_device_runtime_get(dev);  <- Results in `PM_DEVICE_ACTION_TURN_ON` on the device
do things...
pm_device_runtime_put(dev); <- No callback run
k_sleep(K_SECONDS(3));
pm_device_runtime_get(dev); <- Results in `PM_DEVICE_ACTION_TURN_ON` on the device
do things...
pm_device_runtime_put(dev); <- No callback run
```
The device has had `PM_DEVICE_ACTION_TURN_ON` run on it twice, with no physical change in power state. This is not conducive to writing reliable drivers.
IMO this uncontrolled domain can only make sense if the actions are run only when the domain physically changes state.

Reply:
If I understood correctly, `pm_action` shouldn't call `pd_gpio_monitor_update_status`, but check cached variable in data struct, and then return 0 or error. New `pm_action` should like this:
```
switch (action) {
	case PM_DEVICE_ACTION_TURN_ON:
	case PM_DEVICE_ACTION_TURN_OFF:
		return -ENOTSUP;
	case PM_DEVICE_ACTION_RESUME:
		if (!data->isPowered) {
		         rc = -EAGAIN;
                }
		break;
	default:
		break;
}
```
This would still enable support for `PM_DEVICE_RUNTIME`, and don't give any undesirable actions, only giving information about current power domain state.

Reply:
Already updated files, now pm_action only check `is_powered` boolean and then returns proper value, `is_powered` is only changed during gpio callback.